WeakPtr<Page> null dereference crash under EventHandler::handleMousePressEvent

aprotyas opened the pull request:
#### 7d54e81b374f0d872acfe52920a5169ee1d9769e
<pre>
WeakPtr&lt;Page&gt; null dereference crash under EventHandler::handleMousePressEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=296791">https://bugs.webkit.org/show_bug.cgi?id=296791</a>
<a href="https://rdar.apple.com/42023304">rdar://42023304</a>

Reviewed by Richard Robinson.

Recently, we have observed some null dereference crashes under
EventHandler::handleMousePressEvent(), all of the nature:

```
Exception Type:    EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes:   0x0000000000000001, 0x0000000000000008

WebCore::Page::WeakValueType* WTF::WeakPtrImplBase&lt;WTF::DefaultWeakPtrImpl&gt;::get&lt;WebCore::Page&gt;()
  WTF::WeakPtr&lt;WebCore::Page, WTF::DefaultWeakPtrImpl, WTF::RawPtrTraits&lt;WTF::DefaultWeakPtrImpl&gt;&gt;::get() const
    WebCore::Frame::protectedPage() const
      WebCore::EventHandler::handleMousePressEvent(WebCore::PlatformMouseEvent const&amp;)
```

... which indicates that `WeakPtr&lt;Page&gt;` in `EventHandler::m_frame`
is holding on to a nullptr. Instead of unconditionally accessing this
object, this patch makes the codepath less crash prone by introducing a
null check.

No new tests because I was not able to create a reproduction for the
crash yet.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMousePressEvent):

Canonical link: <a href="https://commits.webkit.org/298132@main">https://commits.webkit.org/298132@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/689003135bee2aeb1662f6faac477b1d42412c9f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34069 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24533 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120488 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65037 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dfe1f280-a53b-4110-b375-44a0a9a06655) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42630 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86874 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/59af5abd-5563-46bf-8135-befe0832ec31) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117272 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27631 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67266 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26812 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20798 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64178 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97006 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123698 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41338 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95699 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41715 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98873 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95483 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40632 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37382 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18323 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41218 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46726 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40813 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44119 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42561 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1934,14 +1934,18 @@ HandleUserInputEventResult EventHandler::handleMousePressEvent(const PlatformMou
     }
 
+    RefPtr page = frame->page();
+    if (!page)
```
maybe a debug assertion and/or logging might be useful to figure out the underlying issue eventually?

aprotyas replied:
An assertion won't say anything different than the crash logs, but some logging elsewhere may be informative, true. Let me see what might help!

ddkilzer replied:
A `WeakPtr<>`, by definition, could hold a `nullptr` value at any time, so it should always be checked for a `nullptr` value before being dereferenced.

If we do not expect it to be `nullptr`, we should use a different smart pointer type (maybe `CheckedPtr<>`), although this crash implies that a different smart pointer type would still crash.

Keeping the `nullptr` check and adding logging seems like a good plan if we want to change the `WeakPtr<>` to something else eventually.

aprotyas replied:
Yes, exactly. I agree with the ownership model here (page should be held as a WeakPtr), so a null check is in order but maybe no logging because we don't expect it to always stay non-null.

aprotyas committed:
WeakPtr<Page> null dereference crash under EventHandler::handleMousePressEvent
https://bugs.webkit.org/show_bug.cgi?id=296791
rdar://42023304

Reviewed by Richard Robinson.

Recently, we have observed some null dereference crashes under
EventHandler::handleMousePressEvent(), all of the nature:

```
Exception Type:    EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes:   0x0000000000000001, 0x0000000000000008

WebCore::Page::WeakValueType* WTF::WeakPtrImplBase<WTF::DefaultWeakPtrImpl>::get<WebCore::Page>()
  WTF::WeakPtr<WebCore::Page, WTF::DefaultWeakPtrImpl, WTF::RawPtrTraits<WTF::DefaultWeakPtrImpl>>::get() const
    WebCore::Frame::protectedPage() const
      WebCore::EventHandler::handleMousePressEvent(WebCore::PlatformMouseEvent const&)
```

... which indicates that `WeakPtr<Page>` in `EventHandler::m_frame`
is holding on to a nullptr. Instead of unconditionally accessing this
object, this patch makes the codepath less crash prone by introducing a
null check.

No new tests because I was not able to create a reproduction for the
crash yet.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMousePressEvent):

Canonical link: https://commits.webkit.org/298132@main
